File: js/nanocodex-tools/tools/observation.mjs
export const MAX_OBSERVATION_IMAGE_BYTES = 180_000;
export const OBSERVATION_TIMEOUT_MS = 5_000;
export const OBSERVATION_INTERVAL_MS = 250;

export function observationId(value) {
  if (typeof value !== "string" || !/^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/.test(value)) {
    throw new TypeError("invalid observation identifier");
  }
  return value;
}
function keys(value, allowed) {
  if (!value || typeof value !== "object" || Array.isArray(value)
    || Object.keys(value).some((key) => !allowed.includes(key))) throw new TypeError("invalid observation fields");
}
export function normalizeObservationSurfaces(value) {
  if (!Array.isArray(value) || value.length < 1 || value.length > 8) throw new TypeError("expected 1-8 observation surfaces");
  const ids = new Set();
  return Object.freeze(value.map((surface) => {
    keys(surface, ["id", "name", "kind"]);
    const id = observationId(surface.id);
    if (ids.has(id) || typeof surface.name !== "string" || !surface.name.trim()
      || new TextEncoder().encode(surface.name).length > 128
      || !["desktop", "browser", "phone"].includes(surface.kind)) throw new TypeError("invalid observation surface");
    ids.add(id);
    return Object.freeze({ id, name: surface.name, kind: surface.kind });
  }));
}
export function normalizeObservationProvider(provider, machines) {
  if (provider === undefined) return undefined;
  if (machines.length !== 1 || typeof provider?.capture !== "function") throw new TypeError("observation requires one machine and a capture function");
  return Object.freeze({ surfaces: normalizeObservationSurfaces(provider.surfaces), capture: provider.capture.bind(provider) });
}
export function normalizeObservationFrame(frame) {
  keys(frame, ["captured_at", "width", "height", "mime_type", "data"]);
  if (!Number.isSafeInteger(frame.captured_at) || frame.captured_at < 0
    || ![frame.width, frame.height].every((n) => Number.isInteger(n) && n >= 1 && n <= 8192)
    || !["image/jpeg", "image/png"].includes(frame.mime_type)
    || typeof frame.data !== "string" || frame.data.length === 0
    || frame.data.length > MAX_OBSERVATION_IMAGE_BYTES * 4 / 3
    || !/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(frame.data)) {
    throw new TypeError("invalid observation frame");
  }
  return Object.freeze({ ...frame });
}
export function normalizeObservationResult(result) {
  if (result?.status === "frame") {
    keys(result, ["status", "frame"]);
    return Object.freeze({ status: "frame", frame: normalizeObservationFrame(result.frame) });
  }
  keys(result, ["status", "message"]);
  if (result.status !== "unavailable" || typeof result.message !== "string" || result.message.length > 256) throw new TypeError("invalid observation result");
  return Object.freeze({ status: "unavailable", message: result.message });
}

File: crates/nanocodex-tools/src/attachment/observation.rs
//! Demand-driven screen observations, independent of durable tool execution.

use base64::{Engine as _, engine::general_purpose::STANDARD};
use serde::Serialize;

use super::{AttachmentError, valid_safe_identifier};

/// Maximum decoded image size accepted by every attachment implementation.
pub const MAX_OBSERVATION_IMAGE_BYTES: usize = 180_000;
/// Capture deadline shared with the managed observation broker.
pub const OBSERVATION_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(5);

/// Physical or virtual screen represented by a hand.
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ObservationKind {
    /// Host desktop.
    Desktop,
    /// Browser viewport.
    Browser,
    /// Phone display.
    Phone,
}

/// Immutable, non-secret screen description.
#[derive(Clone, Debug, Serialize)]
pub struct ObservationSurface {
    id: String,
    name: String,
    kind: ObservationKind,
}

impl ObservationSurface {
    /// Creates a bounded screen description.
    ///
    /// # Errors
    /// Rejects invalid identifiers and empty or oversized names.
    pub fn new(
        id: impl Into<String>,
        name: impl Into<String>,
        kind: ObservationKind,
    ) -> Result<Self, AttachmentError> {
        let id = id.into();
        let name = name.into();
        if !valid_safe_identifier(&id, 128) || name.trim().is_empty() || name.len() > 128 {
            return Err(AttachmentError::Catalog(
                "invalid observation surface".into(),
            ));
        }
        Ok(Self { id, name, kind })
    }

    /// Stable screen identifier within this attachment.
    #[must_use]
    pub fn id(&self) -> &str {
        &self.id
    }
}

/// Supported inert image formats; no URLs, SVG, or executable content.
#[derive(Clone, Copy, Debug, Serialize)]
pub enum ObservationImageFormat {
    /// JPEG image.
    #[serde(rename = "image/jpeg")]
    Jpeg,
    /// PNG image.
    #[serde(rename = "image/png")]
    Png,
}

/// One bounded image captured in response to a viewer request.
#[derive(Debug, Serialize)]
pub struct ObservationFrame {
    captured_at: u64,
    width: u32,
    height: u32,
    mime_type: ObservationImageFormat,
    data: String,
}

impl ObservationFrame {
    /// Encodes an image for the cross-language observation protocol.
    ///
    /// # Errors
    /// Rejects oversized or empty images, dimensions, and non-JavaScript timestamps.
    pub fn new(
        captured_at: u64,
        width: u32,
        height: u32,
        format: ObservationImageFormat,
        bytes: &[u8],
    ) -> Result<Self, AttachmentError> {
        if captured_at > 9_007_199_254_740_991
            || !(1..=8192).contains(&width)
            || !(1..=8192).contains(&height)
            || bytes.is_empty()
            || bytes.len() > MAX_OBSERVATION_IMAGE_BYTES
        {
            return Err(AttachmentError::Catalog("invalid observation frame".into()));
        }
        Ok(Self {
            captured_at,
            width,
            height,
            mime_type: format,
            data: STANDARD.encode(bytes),
        })
    }
}

/// Application-owned capture capability. Capture futures must be cancellation
/// safe: dropping one must stop capture and release its resources. The driver
/// permits one capture at a time and drops it on timeout, cancellation, or detach.
#[async_trait::async_trait]
pub trait ObservationProvider: Send + Sync + 'static {
    /// Descriptors snapshotted once when the attachment starts (1–8 unique IDs).
    fn surfaces(&self) -> &[ObservationSurface];
    /// Captures the requested surface. Errors are redacted before transmission.
    async fn capture(&self, surface_id: &str) -> Result<ObservationFrame, AttachmentError>;
}

#[derive(Debug, Serialize)]
#[serde(tag = "status", rename_all = "snake_case")]
pub(crate) enum ObservationResult {
    Frame { frame: ObservationFrame },
    Unavailable { message: &'static str },
}

impl ObservationResult {
    pub(crate) const fn unavailable() -> Self {
        Self::Unavailable {
            message: "Screen capture unavailable",
        }
    }
}
